docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+    // all state necessary to sync.
+    struct GlobalEngineState {
+        // note: no "declined" or "enabled" here - that's invisible to the engine.
+
+        // the result of info/configuration.
+        config: GlobalConfig,
+
+        // The token server token.
+        storage_token: Vec[u8] (??), // not clear on the type here, but whatever.
+
+        // Info about the collections managed by this engine. In most cases there
+        // will be exactly 1.
+        collections: HashMap<String, CollectionState>
+    }
+
+    // The state for a collection. Most engines will manage a single collection, but
+    // there may be many.
+    struct CollectionState {
+        // The info/collections response for this collection.
+        collection_info: CollectionInfo,
+
+        // The current syncID for the engine, as read from meta/global. May be None.
+        // Engines are responsible for persisting this and taking action when it differs.
+        sync_id: Option<SyncGuid>,
+
+        // The keys to be used for this collection. This may be the default key (ie,
+        // used for all collections) or one specific to the collection, but the
+        // engine doesn't need to know that.
+        keys: KeyBundle,
+
+        // Commands the engine should act upon before syncing.
+        commands: Vec<Command>
+    }
+
+     struct Command {
+        ...
+    }
+
+## APIs
+## Sync Manager
+
+    // Register an engine.
+    fn register_engine(name: String, collections: Vec<String>, sync_callback: Fn<...>)
+
+    // Get the list of engines which are declined. There's a bit of confusion here
+    // between an "engine" and a "collection" - eg, assuming history ends up using
+    // 2 collections, the UI will want 1 entry.
+    // Maybe we define this as a "collection" and the app should ignore collections
+    // it doesn't understand?
+    // Will return None if we haven't yet done a first sync so don't know. Will
+    // return an empty vec if we have synced but no engines are declined.
+    fn get_declined() -> Option<Vec<String>>
+
+    // Ditto here - a bit of confusion between "collection" and "engine" here.
+    fn enable_engine(String, bool)
+
+    // TODO: let's think a little more about how desktop uses a single "addons"
+    // pref which controls addons *and* storage.sync, and how a single "Forms"
+    // pref controls 2 discrete engines.
+
+    // Do a sync now. There's a special error code to indicate that something
+    // seems to have gone wrong with auth, which is surfaced to the app, so
+    // it can call on FxA to do magic.
+    fn sync_now(engines: Option<Vec<String>>) -> Result<()>
+
+    // Stop syncing ASAP. Called as the app is shutting down, and possibly
+    // in other circumstances, such as when wifi is lost but the user has
+    // opted in to syncing only on wifi
+    // TBD - does this make sense? If so, how long is the app expected to wait?
+    fn stop()
```

Comment:
I think iOS inverts this a bit and has a "green light", where each engine periodically checks "can I keep syncing?" So I think calling `stop` would set a flag, and the engines would check that periodically as they do work.

Reply:
Yes, that makes sense. Most engines would probably just reuse our existing `InterruptHandle`